Fix invalid WeakRef deference crash in WebCore's Memory Pressure Handler

NathanSolomon733 opened the pull request:
#### 20c26a93818c4b146643e03fcd3c47392bff6f94
<pre>
Fix invalid WeakRef deference crash in WebCore&apos;s Memory Pressure Handler
<a href="https://bugs.webkit.org/show_bug.cgi?id=289117">https://bugs.webkit.org/show_bug.cgi?id=289117</a>
<a href="https://rdar.apple.com/144487022">rdar://144487022</a>

Reviewed by Ryan Reno, Cameron McCormack, and Matthieu Dubet.

We received crash reports of an invalid access to the document
WeakRef while iterating through the allDocuments() map in releaseNoncriticalMemory.
It is not clear what is causing the document to be deleted, perhaps
related to clearing async node deletion list . Although, this should not be possible due
to excluding iframes to be added to the deletion queue and dissallowing Documents
and DocumentFragments to have their immediate children be async deleted.

A fix to this specific crash is to create a list of Ref&lt;Document&gt; from the allDocuments() map,
and iterate through this list.

Change the WeakRef&lt;Document&gt; used to call member functions on the
document into a Ref&lt;Document&gt;, and ensure that the DocumentsMap is not being
modified while clearing the various data structures that the Document holds.

Now, no documents in the map can be deleted while in this function.

* Source/WebCore/page/MemoryRelease.cpp:
(WebCore::releaseNoncriticalMemory):

Canonical link: <a href="https://commits.webkit.org/291644@main">https://commits.webkit.org/291644@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/772153d7510485120e53b2386ea6dbb251c69de3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/93405 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/12965 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/2698 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/98405 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/43931 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/95455 "Passed tests") | [❌ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/13384 "Hash 772153d7 for PR 41879 does not build (failure)") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/21421 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/71356 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28747 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/96407 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/130/builds/13384 "Hash 772153d7 for PR 41879 does not build (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/84477 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/51690 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/130/builds/13384 "Hash 772153d7 for PR 41879 does not build (failure)") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/2107 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/43245 "Built successfully") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/130/builds/13384 "Hash 772153d7 for PR 41879 does not build (failure)") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/2128 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/100437 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/20457 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/14965 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/80375 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/active-view-transition-type-on-non-root.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/20709 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/80395 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79700 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19843 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24243 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/1593 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/13582 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/20441 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/25619 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/20128 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/23588 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/21869 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

heycam reviewed (approved):
s/deference/dereference/ in the commit message title, and there is a stray space before a period in the commit message body.

mdubet commented on Source/WebCore/page/MemoryRelease.cpp:
```diff
@@ -84,7 +84,12 @@ static void releaseNoncriticalMemory(MaintainMemoryCache maintainMemoryCache)
+        return document.get();
+    });
+
+    for (auto& document : protectedDocuments) {
```
Can't we just use `for(Ref document : allDocuments())` ?

rreno replied:
No because if that protected ref is the last ref to the document it will go out of scope at the end of the loop iteration and potentially mutate the allDocumentsMap.

NathanSolomon733 committed:
Fix invalid WeakRef deference crash in WebCore's Memory Pressure Handler
https://bugs.webkit.org/show_bug.cgi?id=289117
rdar://144487022

Reviewed by Ryan Reno, Cameron McCormack, and Matthieu Dubet.

We received crash reports of an invalid access to the document
WeakRef while iterating through the allDocuments() map in releaseNoncriticalMemory.
It is not clear what is causing the document to be deleted, perhaps
related to clearing async node deletion list . Although, this should not be possible due
to excluding iframes to be added to the deletion queue and dissallowing Documents
and DocumentFragments to have their immediate children be async deleted.

A fix to this specific crash is to create a list of Ref<Document> from the allDocuments() map,
and iterate through this list.

Change the WeakRef<Document> used to call member functions on the
document into a Ref<Document>, and ensure that the DocumentsMap is not being
modified while clearing the various data structures that the Document holds.

Now, no documents in the map can be deleted while in this function.

* Source/WebCore/page/MemoryRelease.cpp:
(WebCore::releaseNoncriticalMemory):

Canonical link: https://commits.webkit.org/291644@main
